buildSrc/src/main/kotlin/conventions/optional.gradle.kts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package conventions
+
+// Provides the optional dependencies support eg: optionalApi, optionalImplementation
+plugins {
+    id("java-library")
+    id("maven-publish")
+}
+
+java { registerFeature("optional") { usingSourceSet(sourceSets["main"]) } }
```

Comment:
Where is this "optional" feature actually defined?  It's not clear to me how this works :)

Reply:
We're registering the feature here. See: https://docs.gradle.org/current/userguide/feature_variants.html for more information.